### PR TITLE
[WIP] feat(GUI): WiFi methods integration with dbus NetworkManager

### PR DESCRIPTION
**Pull request by thundron**
Integrate functional logic (connect, disconnect, list network(s)) inside WiFi settings modal made in #2324 
Add on-screen keyboard to allow passphrase configuration (ref: #2253 )
The WiFi button is hidden behind the `ETCHER_ENABLE_WIFI_DIALOG` env variable.
The WiFi service endpoint is composed through the env variables:
- `ETCHER_WIFI_SERVICE_IP`: defaults to `localhost`
- `ETCHER_WIFI_SERVICE_PORT`: defaults to `1337`

**TODO:**
- [x] Discuss error handling in product call
- [x] Refactor keyboard to span full screen

Closes: #2252 , #2253 
Change-type: patch
Changelog-Entry: WiFi settings modal and on-screen keyboard for Etcher Pro

Old branch description for reference:

> feat(GUI):
> 
>  wifi settings modal for etcher pro
> 
> We add a new component for configuring WiFi on Pro devices, written
> in Rendition and React with styled-components. It is hidden behind an
> environment variable `ETCHER_ENABLE_WIFI_DIALOG`.
> 
> Closes: #2252
> Change-Type: patch
> Changelog-Entry: WiFi settings modal for Etcher Pro.

### CODE REVIEW

File: lib/gui/app/components/wifi/controllers/wifi.js [hunk 1]
```diff
@@ -0,0 +1,32 @@
+
+/*
+ * Copyright 2016 resin.io
```
Comment (lurch) [inline lib/gui/app/components/wifi/controllers/wifi.js]: 2016?? :wink: 
And some of the other files added in this PR seem to be missing the appropriate copyright header?

**Comment by lurch** [comment]
...and Travis seems to be complaining because you've modified `package.json`, but not pushed the corresponding changes to `npm-shrinkwrap.json` ?

File: lib/gui/app/components/wifi/react/keyboard/index.jsx [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+class KBInput extends React.PureComponent {
+  componentDidMount () {
+    this.keyboard.interface.keyaction.enter = (base) => {
+      return this.keyboard.interface.keyaction.accept(base)
+    }
+  }
+
+  render () {
+    console.log('value', this.props.value, typeof this.props.value)
+    console.log('type', this.props.type, typeof this.props.type)
```
Comment (jhermsmeier) [inline lib/gui/app/components/wifi/react/keyboard/index.jsx]: Leftover debug statements? If you'd like to keep them around, I'd suggest using the `debug` module (see elsewhere in Etcher)

File: lib/gui/app/components/wifi/react/keyboard/react-virtual-keyboard.custom.jsx [hunk 3]
```diff
@@ -0,0 +1,160 @@
+'use strict'
+
+const React = require('react')
+const PropTypes = require('prop-types')
+const jQuery = require('jquery')
+const keyboard = require('virtual-keyboard/dist/js/jquery.keyboard.js')
```
Comment (jhermsmeier) [inline lib/gui/app/components/wifi/react/keyboard/react-virtual-keyboard.custom.jsx]: Is this mixing React & jQuery? What's the reason for this – is this a requirement of the virtual-keyboard module?

File: package.json [hunk 4]
```diff
@@ -40,7 +40,9 @@
     "fsevents"
   ],
   "dependencies": {
-    "@fortawesome/fontawesome-free-webfonts": "^1.0.9",
+    "@fortawesome/fontawesome": "1.1.8",
+    "@fortawesome/fontawesome-free-solid": "5.0.13",
```
Comment (jhermsmeier) [inline package.json]: Why are these needed (what requires these)? We've previously moved off of them due to performance issues with every `render()` re-creating inline SVGs

File: lib/gui/app/components/wifi/react/keyboard/react-virtual-keyboard.custom.jsx (hunk 3 above)
Comment (thundron) [reply lib/gui/app/components/wifi/react/keyboard/react-virtual-keyboard.custom.jsx]: I deleted this file, it was to circumvent that stupid `type` error I was getting

File: lib/gui/app/components/wifi/react/keyboard/index.jsx (hunk 2 above)
Comment (thundron) [reply lib/gui/app/components/wifi/react/keyboard/index.jsx]: Yep, it's still a WIP so I'm gonna sanitize all console.logs and maybe introduce a bit of analytics too

File: package.json (hunk 4 above)
Comment (thundron) [reply package.json]: I dunno, it gave me errors back when I started developing on the branch. I think I can get rid of them.
Should I just keep `fontawesome-free-webfonts` or delete that as well?

Comment (jhermsmeier) [reply package.json]: > Should I just keep fontawesome-free-webfonts or delete that as well?

Keep keep, that's used by the file-picker ;)

File: lib/gui/app/components/wifi/react/actions.js [hunk 5]
```diff
@@ -0,0 +1,199 @@
...
+ */
+
+'use strict'
+
+const _ = require('lodash')
+
+const wifiServiceIP = process.env.ETCHER_WIFI_SERVICE_IP || 'localhost'
+const wifiServicePORT = process.env.ETCHER_WIFI_SERVICE_PORT || '1337'
+
+const wifiServiceEndpoint = `http://${wifiServiceIP}:${wifiServicePORT}/wifi`
```
Comment (jhermsmeier) [inline lib/gui/app/components/wifi/react/actions.js]: Can we use the `url` module to construct the endpoint?

```js

File: lib/gui/app/components/wifi/react/actions.js [hunk 6]
```diff
@@ -0,0 +1,199 @@
...
+        }
+        throw res.json().then((err) => {
+          throw err
+        })
+      })
+      .then((res) => {
+        return res.value
+      })
+      .catch((err) => {
+        throw err
```
Comment (jhermsmeier) [inline lib/gui/app/components/wifi/react/actions.js]: Catch & re-throw can be removed

File: lib/gui/app/components/wifi/react/keyboard/style.scss [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+
+	.ui-keyboard-button {
+		background-color: #3e3e3e;
+		color: whitesmoke;
+    padding: 0.5em 1.5em;
+    margin: 0.2em;
+    border: 0;
+    border-radius: 0.1em;
+    font-size: 1.75em;
+    border-top: 1px solid #4e4e4e;
```
Comment (jhermsmeier) [inline lib/gui/app/components/wifi/react/keyboard/style.scss]: Indentation needs fixing here

File: lib/gui/app/components/wifi/react/keyboard/style.scss [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+	}
+
+	.ui-keyboard-space {
+		width: 40%;
+  }
```
Comment (jhermsmeier) [inline lib/gui/app/components/wifi/react/keyboard/style.scss]: Indentation here as well

File: lib/gui/app/components/wifi/react/wifi.jsx [hunk 9]
```diff
@@ -0,0 +1,584 @@
...
+const KBInput = require('./keyboard/index.jsx')
+
+/**
+ * @summary Font awesome icon constants
+ */
+// const faWifiHTML = fontAwesome.icon(faWifi).html.join('\n')
+// const faLockHTML = fontAwesome.icon(faLock).html.join('\n')
+// const faWrenchHTML = fontAwesome.icon(faWrench).html.join('\n')
+// const faAngleLeftHTML = fontAwesome.icon(faAngleLeft).html.join('\n')
+// const faTimesHTML = fontAwesome.icon(faTimes).html.join('\n')
```
Comment (jhermsmeier) [inline lib/gui/app/components/wifi/react/wifi.jsx]: Looks like some leftover comments that can be removed?

File: lib/gui/app/components/wifi/react/wifi.jsx [hunk 10]
```diff
@@ -0,0 +1,584 @@
...
+      <span
+        style={ { opacity: isOpen ? '0' : '1' } }
+        className="fas fa-lock" />
+        {/* dangerouslySetInnerHTML={ { __html: faLockHTML } } /> */}
+      { labelElem }
+      { props.selected ? configureButton : null }
+    </div>
+  )
+})
+const Connection = ConnectionUnstyled `
```
Comment (jhermsmeier) [inline lib/gui/app/components/wifi/react/wifi.jsx]: Can we add a newline between declarations here, for readability?

File: lib/gui/app/components/wifi/react/wifi.jsx [hunk 11]
```diff
@@ -0,0 +1,584 @@
...
+    }
+  }
+
+  configureNetwork (network) {
+    return () => {
+      this.selectedEdit = network
+      this.browse('configure')
+      // this.setState({
+      //   page: 'configure'
+      // })
```
Comment (jhermsmeier) [inline lib/gui/app/components/wifi/react/wifi.jsx]: Comment ✂️

**Comment by jhermsmeier** [comment]
CI is failing due to mangled shrinkwrap; looks like there are some leftovers from removing dependencies. You can run that check locally with `make sanity-checks`

File: lib/gui/app/components/wifi/react/keyboard/style.scss (hunk 7 above)
Comment (thundron) [reply lib/gui/app/components/wifi/react/keyboard/style.scss]: This should be fixed, although didn't see any error/warning on my editor... They're all the same level. Gonna retry de-tab/re-tabbing them

**Comment by jhermsmeier** [comment]
Getting a warning about prop-types:

```
Warning: Failed prop type: Wifi: prop type `close` is invalid; it must be a function, usually from the `prop-types` package, but received `undefined`.
```

And the blue close button doesn't work (which looks like it's due to above)

File: lib/gui/app/components/wifi/react/actions.js [hunk 12]
```diff
@@ -21,7 +21,7 @@ const url = require('url')
 
 const wifiServiceEndpoint = url.format({
   protocol: 'http',
-  hostname: process.env.ETCHER_WIFI_SERVICE_IP || 'localhost',
+  hostname: process.env.ETCHER_WIFI_SERVICE_IP || '10.211.55.3' || 'localhost',
```
Comment (jhermsmeier) [inline lib/gui/app/components/wifi/react/actions.js]: I don't think that IP is supposed to be in here ;)

File: lib/gui/app/components/wifi/react/wifi.jsx [hunk 13]
```diff
@@ -453,8 +453,9 @@ class Wifi extends React.PureComponent {
    */
   handleError (err) {
     console.error(err)
+    const message = (!_.isEmpty(err.data) && _.head(err.data)) || err.message
     this.setState({
-      wifiSettingsError: err
+      wifiSettingsError: { message }
```
Comment (jhermsmeier) [inline lib/gui/app/components/wifi/react/wifi.jsx]: `new Error( message )` please ;P

File: lib/gui/app/components/wifi/react/actions.js (hunk 12 above)
Comment (thundron) [reply lib/gui/app/components/wifi/react/actions.js]: Definetly not. Fixed

File: lib/gui/app/components/wifi/react/wifi.jsx (hunk 13 above)
Comment (thundron) [reply lib/gui/app/components/wifi/react/wifi.jsx]: Yeah, and I should definitely keep this in mind
